[HDR] Ensure HDR layers and IOSurfaces are created only when they contain HDR contents

shallawa opened the pull request:
#### 1ffb2f06c35a2fa0af0654a8c9de49197b19fa07
<pre>
[HDR] Ensure HDR layers and IOSurfaces are created only when they contain HDR contents
<a href="https://bugs.webkit.org/show_bug.cgi?id=282298">https://bugs.webkit.org/show_bug.cgi?id=282298</a>
<a href="https://rdar.apple.com/138877325">rdar://138877325</a>

Reviewed by Simon Fraser.

Each layer will decide whether it needs to draw HDR content or not. The renderer
of the RenderLayer and its sub-tree will be inspected for any HDR image or canvas.

RenderLayer::isReplacedElementWithHDR() will return true if the renderer of the
layer itself needs to draw HDR content. The image of the ImageDocument will be
handled as a special case.

PaintedContentsInfo::isReplacedElementWithHDR() will return whether the renderer
of the layer itself will need to draw HDR content or not.

All the virtual methods hdrForImagesEnabled() will be deleted and replaced by
drawsHDRContent(). drawsHDRContent() will be set to true only if the layer will
need to draw HDR contents and settings.hdrForImagesEnabled() is true.

Since determineNonLayerDescendantsPaintedContent() traverses the render tree of
layer to set hasPaintedContent, it will be used also to set hasPaintedHDRContent
of PaintedContentRequest.

Since RenderLayerBacking::updateDrawsContent() sets GraphicsLayer::setDrawsContent
from the renderer of the layer itself, this function will be used also to set
GraphicsLayer::setDrawsHDRContent() by calling
PaintedContentsInfo::isReplacedElementWithHDR().

PlatformCALayer::contentsFormatForLayer() will call GraphicsLayerCA::drawsHDRContent()
instead of calling hdrForImagesEnabled() to know whether to return RGBA16F for
this layer or not.

For testing support the internal API setScreenContentsFormatsForTesting() is added
to force the screen supported ContentsFormats to certain values.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-basic-canvas.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-basic-canvas-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-canvas-expected.txt: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::isHDR const):
(WebCore::CanvasRenderingContext::drawsHDRContent const): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::canDrawHDRContents const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageOverlayController.cpp:
(WebCore::PageOverlayController::updateSettingsForLayer):
* Source/WebCore/platform/PlatformScreen.cpp:
(WebCore::setScreenContentsFormatsForTesting):
(WebCore::screenContentsFormatsForTesting):
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/ScreenProperties.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::setDrawsHDRContent):
(WebCore::GraphicsLayer::dumpProperties const):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::drawsHDRContent const):
(WebCore::GraphicsLayer::setHDRForImagesEnabled): Deleted.
(WebCore::GraphicsLayer::hdrForImagesEnabled const): Deleted.
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::hdrForImagesEnabled const): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setDrawsHDRContent):
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::drawRepaintIndicator):
(WebCore::PlatformCALayer::contentsFormatForLayer):
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::drawsHDRContent const):
(WebCore::PlatformCALayerClient::hdrForImagesEnabled const): Deleted.
* Source/WebCore/platform/ios/PlatformScreenIOS.mm:
(WebCore::screenSupportsExtendedColor):
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::screenSupportsExtendedColor):
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::PaintedContentsInfo):
(WebCore::PaintedContentsInfo::isPaintsContentSatisfied const):
(WebCore::PaintedContentsInfo::paintsHDRContent):
(WebCore::PaintedContentsInfo::isContentsTypeSatisfied const):
(WebCore::PaintedContentsInfo::isSimpleContainer):
(WebCore::PaintedContentsInfo::isDirectlyCompositedImage):
(WebCore::PaintedContentsInfo::isUnscaledBitmapOnly):
(WebCore::PaintedContentsInfo::isReplacedElementWithHDR):
(WebCore::PaintedContentsInfo::determinePaintsContent):
(WebCore::PaintedContentsInfo::determineContentsType):
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::determineNonCompositedLayerDescendantsPaintedContent const):
(WebCore::RenderLayerBacking::isReplacedElementWithHDR const):
(WebCore::PaintedContentsInfo::contentsTypeDetermination): Deleted.
(WebCore::RenderLayerBacking::hdrForImagesEnabled const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::hdrForImagesEnabled const): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setScreenContentsFormatsForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::hdrForImagesEnabled const): Deleted.

Canonical link: <a href="https://commits.webkit.org/290503@main">https://commits.webkit.org/290503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cde3dbea1f239991d2d9812d1c74a663367302cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90237 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9766 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45160 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95239 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41013 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92289 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10154 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18081 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69472 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27069 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93238 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49832 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7511 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40144 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37294 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97072 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17425 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78469 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17682 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77689 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77675 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19183 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22138 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20740 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10684 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17435 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17176 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18960 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -3253,6 +3274,13 @@ void GraphicsLayerCA::updateReplicatedLayers()
+#if HAVE(HDR_SUPPORT)
+void GraphicsLayerCA::updateDrawsHDRContent()
+{
+    m_layer->setContentsFormat(screenContentsFormat(nullptr, this));
```
`screenContentsFormat` is a misnomer; it's not the contents format used by the "screen", it just the contents format we've chosen for HDR.

I'd still like us to separate "is showing HDR" from "contents format".

shallawa replied:
I created https://github.com/WebKit/WebKit/pull/40225 to address this issue.

smfr commented on Source/WebCore/rendering/RenderHTMLCanvas.cpp:
```diff
@@ -71,6 +71,18 @@ bool RenderHTMLCanvas::requiresLayer() const
     return canvasCompositingStrategy(*this) != CanvasPaintedToEnclosingLayer;
 }
 
+void RenderHTMLCanvas::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle)
```
I'm not sure this is right; a style change doesn't mean that the context state changed, and calling `contentChanged` on every style change seems wrong.

Why doesn't the call to `canvasRenderer->contentChanged(ContentChangeType::Canvas)`  in `HTMLCanvasElement::reset()` already take care of this?

shallawa replied:
You right. I removed this code and I used `RenderLayerBacking::updateDrawsContent()` to check whether the renderer of `RenderLayer` itself drawsHDRContent or not and `GraphicsLayer::setDrawsHDRContent()` accordingly.

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -120,12 +120,31 @@ enum ShouldApplyRootOffsetToFragments {
-    False,
-    True,
-    Undetermined
+struct RequestState {
```
Could you explain in the changelog why these changes were needed?

shallawa replied:
This change was removed.

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -225,6 +238,18 @@ PaintedContentsInfo::ContentsTypeDetermination PaintedContentsInfo::contentsType
+    if (m_hdrContent != RequestState::Unknown)
+        return m_hdrContent;
+
+    RenderLayer::PaintedContentRequest contentRequest;
```
By making a new `PaintedContentRequest` here, you're missing the opportunity to share the same tree traversal that `PaintedContentsInfo::paintsContentDetermination() does. Use the existing `PaintedContentRequest` if you can.

shallawa replied:
I created https://github.com/WebKit/WebKit/pull/40360 to address this issue.

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -2981,6 +3017,51 @@ bool RenderLayerBacking::paintsBoxDecorations() const
+        return true;
+    }
+
+    for (auto& child : descendantsOfType<RenderReplaced>(renderer())) {
```
This code should be integrated with `RenderLayerBacking::paintsContent` so we only do one layer tree traversal;

shallawa replied:
Fixed.

cdumez commented on Source/WebCore/platform/graphics/GraphicsLayer.cpp:
```diff
@@ -197,6 +197,9 @@ GraphicsLayer::GraphicsLayer(Type type, GraphicsLayerClient& layerClient)
     , m_masksToBounds(false)
     , m_drawsContent(false)
+#if HAVE(HDR_SUPPORT)
+    , m_drawsHDRContent(false)
```
nit: would be nice to inline in the header. As far as I know this works for bitfields with modern C++.

shallawa replied:
Fixed.

cdumez commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -120,12 +120,31 @@ enum ShouldApplyRootOffsetToFragments {
-    True,
-    Undetermined
+struct RequestState {
+    enum State {
```
Going from `enum class` to old style `enum` seems like a step in the wrong direction?

shallawa replied:
This change was removed.

cdumez commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -120,12 +120,31 @@ enum ShouldApplyRootOffsetToFragments {
+
+    constexpr RequestState(State state)
+    {
+        this->state = state;
```
Should use the initialization list.

shallawa replied:
This change was removed.

cdumez commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -120,12 +120,31 @@ enum ShouldApplyRootOffsetToFragments {
+
+    constexpr RequestState(bool state)
+    {
+        this->state = state ? True : False;
```
Should use the initialization list.

shallawa replied:
This change was removed.

cdumez commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -1951,6 +1979,14 @@ void RenderLayerBacking::updateDrawsContent(PaintedContentsInfo& contentsInfo)
 }
 
+#if HAVE(HDR_SUPPORT)
+void RenderLayerBacking::updateDrawsHDRContent(PaintedContentsInfo& contentsInfo)
```
Can the parameter with `const`?

shallawa replied:
This change was removed.

cdumez commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -2981,6 +3017,51 @@ bool RenderLayerBacking::paintsBoxDecorations() const
+    if (!settings.hdrForImagesEnabled() && !settings.canvasPixelFormatEnabled())
+        return false;
+
+    auto hasPaintedHDRContent = [&settings = settings] (RenderElement& renderer) -> bool {
```
Likely doesn't need `-> bool`

shallawa replied:
This is fixed in https://github.com/WebKit/WebKit/pull/40360.

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -2288,6 +2304,11 @@ void GraphicsLayerCA::commitLayerChangesAfterSublayers(CommitState& commitState)
 
+#if HAVE(HDR_SUPPORT)
+    if (m_uncommittedChanges & DrawsHDRContentChanged)
+        updateDrawsHDRContent();
```
Is there a reason this is in `commitLayerChangesAfterSublayers` rather than before sublayers? I don't think it should be in the "after".

shallawa replied:
This code was removed. Calling `updateDrawsHDRContent()` is only called in `commitLayerChangesBeforeSublayers()`

smfr commented on LayoutTests/fast/hdr/hdr-basic-canvas-layer-dump-expected.txt:
```diff
@@ -0,0 +1,56 @@
+
+(GraphicsLayer
```
This test might be better under `compositing/hdr` since it dumps GraphicsLayers.

shallawa replied:
This test was moved to `compositing/hdr/basic-hdr-canvas.html`

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.cpp:
```diff
@@ -444,6 +444,14 @@ void GraphicsLayer::setDrawsContent(bool b)
+#if HAVE(HDR_SUPPORT)
+void GraphicsLayer::setDrawsHDRContent(bool b)
+{
+    ASSERT_IMPLIES(m_type == Type::Structural, false);
```
Why not `ASSERT(m_type != Type::Structural)` or `ASSERT_IMPLIES(m_type == Type::Structural, !b)` ?

shallawa replied:
This assertion was changed to `ASSERT(m_type != Type::Structural)`. I just copied it from `GraphicsLayer::setDrawsContent()`. I changed this assertion only here but I will clean the rest of assertions in a separate PR.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5774,41 +5775,38 @@ static bool hasVisibleBoxDecorationsOrBackground(const RenderElement& renderer)
-#endif
+    if (CheckedPtr imageRenderer = dynamicDowncast<RenderImage>(renderer)) {
+        if (auto* cachedImage = imageRenderer->cachedImage()) {
+            if (cachedImage->drawsHDRContent())
```
It's odd to an an image if it "draws HDR content"; the right question would be "isHDR"? Same with canvas.

shallawa replied:
This code was moved to a separate function named `isReplacedElementWithHDR()`. `CachedImage::drawsHDRContent()` itself was renamed `CachedImage::isHDR()`.

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -168,7 +169,16 @@ class PaintedContentsInfo {
-        determinePaintsContent();
+        if (m_hdrContent == RequestState::Unknown)
+            determinePaintsContent();
+        return m_hdrContent == RequestState::True || m_hdrContent == RequestState::Undetermined;
```
Is it right to say that we need HDR if the state is "Undetermined"? That means that every layer with a huge layer tree will use HDR.

shallawa replied:
Checking `RequestState::Undetermined` was removed. In fact `m_hdrContent` is never set to `Undetermined`.

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -593,6 +593,10 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
     void determineNonLayerDescendantsPaintedContent(PaintedContentRequest&) const;
+#if HAVE(HDR_SUPPORT)
+    // True of if renderer itself draws HDR content, no traversal is done.
+    bool drawsHDRContent() const;
```
Please use consistent terminology with `paintsHDRContent` / `paintHDRContentOnly`. I think this is the "only" version? A good name would make the comment unnecessary. This is just for replaced elements (image, canvas etc), so you could call it `isReplacedElementWithHDR` or something.

shallawa replied:
This function was renamed `isReplacedElementWithHDR()`.

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -168,7 +169,16 @@ class PaintedContentsInfo {
+    }
+
+    void determinePaintsHDRContent();
+    bool paintHDRContentOnly()
```
The naming of `determinePaintsHDRContent` needs to match the `isReplacedElementWithHDR` naming so that the "only" here is clear. I would avoid "only", because an HDR image can have a non-HDR CSS border.

shallawa replied:
This two functions were removed. I added calculating  whether `PaintedContentsInfo` is `isReplacedElementWithHDR` to `PaintedContentsInfo::determineContentsType()`

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -3294,6 +3320,13 @@ bool RenderLayerBacking::isUnscaledBitmapOnly() const
 }
 
+#if HAVE(HDR_SUPPORT)
+bool RenderLayerBacking::drawsHDRContent() const
```
Here I don't know if this is the "only" version or not.

shallawa replied:
This function was renamed `RenderLayerBacking::isReplacedElementWithHDR()`.

smfr commented on Source/WebCore/testing/Internals.idl:
```diff
@@ -469,6 +469,18 @@ enum RenderingMode {
     "Accelerated"
 };
 
+#if defined(HAVE_HDR_SUPPORT)
```
I don't think you should `#ifdef` in this IDL file. Tests should run on all platforms without throwing exceptions, but may just behave like HDR is not available.

shallawa committed:
[HDR] Ensure HDR layers and IOSurfaces are created only when they contain HDR contents
https://bugs.webkit.org/show_bug.cgi?id=282298
rdar://138877325

Reviewed by Simon Fraser.

Each layer will decide whether it needs to draw HDR content or not. The renderer
of the RenderLayer and its sub-tree will be inspected for any HDR image or canvas.

RenderLayer::isReplacedElementWithHDR() will return true if the renderer of the
layer itself needs to draw HDR content. The image of the ImageDocument will be
handled as a special case.

PaintedContentsInfo::isReplacedElementWithHDR() will return whether the renderer
of the layer itself will need to draw HDR content or not.

All the virtual methods hdrForImagesEnabled() will be deleted and replaced by
drawsHDRContent(). drawsHDRContent() will be set to true only if the layer will
need to draw HDR contents and settings.hdrForImagesEnabled() is true.

Since determineNonLayerDescendantsPaintedContent() traverses the render tree of
layer to set hasPaintedContent, it will be used also to set hasPaintedHDRContent
of PaintedContentRequest.

Since RenderLayerBacking::updateDrawsContent() sets GraphicsLayer::setDrawsContent
from the renderer of the layer itself, this function will be used also to set
GraphicsLayer::setDrawsHDRContent() by calling
PaintedContentsInfo::isReplacedElementWithHDR().

PlatformCALayer::contentsFormatForLayer() will call GraphicsLayerCA::drawsHDRContent()
instead of calling hdrForImagesEnabled() to know whether to return RGBA16F for
this layer or not.

For testing support the internal API setScreenContentsFormatsForTesting() is added
to force the screen supported ContentsFormats to certain values.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/hdr/hdr-basic-canvas.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/compositing/hdr/hdr-basic-canvas-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/compositing/hdr/hdr-basic-canvas-expected.txt: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::isHDR const):
(WebCore::CanvasRenderingContext::drawsHDRContent const): Deleted.
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::canDrawHDRContents const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageOverlayController.cpp:
(WebCore::PageOverlayController::updateSettingsForLayer):
* Source/WebCore/platform/PlatformScreen.cpp:
(WebCore::setScreenContentsFormatsForTesting):
(WebCore::screenContentsFormatsForTesting):
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/ScreenProperties.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::setDrawsHDRContent):
(WebCore::GraphicsLayer::dumpProperties const):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::drawsHDRContent const):
(WebCore::GraphicsLayer::setHDRForImagesEnabled): Deleted.
(WebCore::GraphicsLayer::hdrForImagesEnabled const): Deleted.
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::hdrForImagesEnabled const): Deleted.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setDrawsHDRContent):
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::drawRepaintIndicator):
(WebCore::PlatformCALayer::contentsFormatForLayer):
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::drawsHDRContent const):
(WebCore::PlatformCALayerClient::hdrForImagesEnabled const): Deleted.
* Source/WebCore/platform/ios/PlatformScreenIOS.mm:
(WebCore::screenSupportsExtendedColor):
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::screenSupportsExtendedColor):
(WebCore::screenSupportsHighDynamicRange):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::PaintedContentRequest::PaintedContentRequest):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::PaintedContentsInfo::PaintedContentsInfo):
(WebCore::PaintedContentsInfo::isPaintsContentSatisfied const):
(WebCore::PaintedContentsInfo::paintsHDRContent):
(WebCore::PaintedContentsInfo::isContentsTypeSatisfied const):
(WebCore::PaintedContentsInfo::isSimpleContainer):
(WebCore::PaintedContentsInfo::isDirectlyCompositedImage):
(WebCore::PaintedContentsInfo::isUnscaledBitmapOnly):
(WebCore::PaintedContentsInfo::isReplacedElementWithHDR):
(WebCore::PaintedContentsInfo::determinePaintsContent):
(WebCore::PaintedContentsInfo::determineContentsType):
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::determineNonCompositedLayerDescendantsPaintedContent const):
(WebCore::RenderLayerBacking::isReplacedElementWithHDR const):
(WebCore::PaintedContentsInfo::contentsTypeDetermination): Deleted.
(WebCore::RenderLayerBacking::hdrForImagesEnabled const): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::hdrForImagesEnabled const): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setScreenContentsFormatsForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::hdrForImagesEnabled const): Deleted.

Canonical link: https://commits.webkit.org/290503@main
